File: Engine/Android/core/src/test/kotlin/despia/engine/ComposeConformanceTest.kt
package despia.engine

import java.io.File
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The composer conformance runner - executes OpenSource/Conformance/compose/result.json through
 * THIS runtime's ComposeCore folds (parity F13). The TS twin (@despia/kernel compose-core.ts,
 * compose-conformance.test.ts) and the Swift reference (ComposeConformance) run the SAME file,
 * so `unknown` cannot quietly become `sent` on the renderer that never learns what the user did.
 *
 * Missing corpus = loud failure, and so is an empty section.
 */
class ComposeConformanceTest {

    private fun corpusFile(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/compose/result.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/compose/result.json not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun root(): Map<String, Any?> {
        val root = json(corpusFile().readText()).foundationValue as? Map<String, Any?>
            ?: error("result.json: not a JSON object")
        assertEquals(1, (root["version"] as? Number)?.toInt(), "result.json version")
        return root
    }

    @Suppress("UNCHECKED_CAST")
    private fun section(name: String): List<Map<String, Any?>> {
        val raw = root()[name]
        val rows = (raw as? List<Any?>) ?: ((raw as? Map<String, Any?>)?.get("cases") as? List<Any?>)
        val cases = rows?.map { it as Map<String, Any?> }
        assertTrue(cases != null && cases.isNotEmpty(), "result.json: $name must be a non-empty case array")
        return cases!!
    }

    private fun int(value: Any?): Int = (value as Number).toInt()

    @Test
    @Suppress("UNCHECKED_CAST")
    fun vocabularyAgreesWithCorpus() {
        assertEquals((root()["vocabulary"] as List<Any?>).map { it as String }, ComposeCore.RESULTS)
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun resultFidelityAgreesWithCorpus() {
        val fidelity = (root()["resultFidelity"] as Map<String, Any?>).filterKeys { !it.startsWith("_") }
        assertTrue(fidelity.isNotEmpty(), "result.json: resultFidelity names no renderers")
        for ((renderer, actions) in fidelity) {
            for ((action, results) in (actions as Map<String, Any?>)) {
                assertEquals((results as List<Any?>).map { it as String },
                             ComposeCore.RESULT_FIDELITY[renderer]?.get(action),
                             "resultFidelity: $renderer.$action")
            }
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun permissionSurfaceIsEmptyOnEveryAction() {
        val surface = root()["permissionSurface"] as Map<String, Any?>
        val declared = surface.filterKeys { !it.startsWith("_") }.filterValues { it is String }
        assertTrue(declared.isNotEmpty(), "result.json: permissionSurface names no actions")
        for ((action, grant) in declared) {
            assertEquals("none", grant, "permissionSurface: $action must need no grant")
            assertEquals("none", ComposeCore.PERMISSION_SURFACE[action], "permissionSurface: $action")
        }
        assertEquals((surface["forbiddenPermissions"] as List<Any?>).map { it as String },
                     ComposeCore.FORBIDDEN_PERMISSIONS)
        for (permission in ComposeCore.FORBIDDEN_PERMISSIONS) {
            assertTrue(permission !in ComposeCore.PERMISSION_SURFACE.values,
                       "$permission must never appear in the permission surface")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun resultLadderAgreesWithCorpus() {
        val fidelity = root()["resultFidelity"] as Map<String, Any?>
        for (case in section("cases")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>
            val renderer = given["renderer"] as String
            val action = given["action"] as String

            val outcome = ComposeCore.outcome(
                renderer = renderer,
                composerResult = given["composerResult"] as? String,
                launched = given["launched"] as? Boolean ?: true,
                isHtml = given["isHtml"] == true,
            )

            val expectedError = expect["error"] as? String
            if (expectedError != null) {
                assertEquals(expectedError, outcome.error, "$name: error")
                continue
            }
            assertNull(outcome.error, "$name: error")
            assertEquals(expect["result"], outcome.result, "$name: result")
            if (expect.containsKey("isHtml")) {
                assertEquals(expect["isHtml"] as Boolean?, outcome.isHtml, "$name: isHtml")
            }
            val allowed = ((fidelity[renderer] as Map<String, Any?>)[action] as List<Any?>).map { it as String }
            assertTrue(outcome.result in allowed,
                       "$name: ${outcome.result} is outside $renderer.$action's fidelity list")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun capabilitiesAgreeWithCorpus() {
        for (case in section("capabilities")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>
            val capabilities = ComposeCore.capabilities(
                renderer = given["renderer"] as String,
                canText = given["canText"] == true,
                canMail = given["canMail"] == true,
                smsResolver = given["smsResolver"] as? String,
                mailResolver = given["mailResolver"] as? String,
            )
            assertEquals(expect["sms"], capabilities.sms, "$name: sms")
            assertEquals(expect["mail"], capabilities.mail, "$name: mail")
            if (expect.containsKey("defaultMailClient")) {
                assertEquals(expect["defaultMailClient"] as String?, capabilities.defaultMailClient,
                             "$name: defaultMailClient")
            }
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun recipientCapAgreesWithCorpus() {
        for (case in section("recipients")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>
            val count = if (given.containsKey("count")) int(given["count"])
                        else int(given["to"] ?: 0) + int(given["cc"] ?: 0) + int(given["bcc"] ?: 0)
            val decision = ComposeCore.recipientDecision(count, int(given["cap"]))
            assertEquals(expect["runs"], decision.runs, "$name: runs")
            assertEquals(expect["error"] as String?, decision.error, "$name: error")
        }
    }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun attachmentRuleAgreesWithCorpus() {
        for (case in section("attachments")) {
            val name = case["name"] as? String ?: "<unnamed>"
            val given = case["given"] as Map<String, Any?>
            val expect = case["expect"] as Map<String, Any?>
            val decision = ComposeCore.attachmentDecision(
                renderer = given["renderer"] as String,
                path = given["path"] as? String,
                insideRoots = given["insideRoots"] as? Boolean,
                exists = given["exists"] as? Boolean,
            )
            assertEquals(expect["runs"], decision.runs, "$name: runs")
            assertEquals(expect["error"] as String?, decision.error, "$name: error")
        }
    }
}

File: Web/packages/kernel/test/compose-conformance.test.ts
//
//  compose-conformance.test.ts — the SHARED composer corpus through the TS kernel
//  (OpenSource/Conformance/compose/result.json). The Kotlin twin (:core
//  ComposeConformanceTest) and the Swift reference (ComposeConformance, record lane) run the
//  SAME file, so `unknown` cannot quietly become `sent` on the renderer that never learns what
//  the user did.
//
//  Missing corpus = loud failure, and so is an empty section.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  COMPOSE_RESULTS, COMPOSE_RESULT_FIDELITY, COMPOSE_PERMISSION_SURFACE,
  COMPOSE_FORBIDDEN_PERMISSIONS, composeOutcome, composeCapabilities,
  composeRecipientDecision, composeAttachmentDecision,
} from "../src/compose-core.ts";

type Json = { [key: string]: any };

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/compose");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("OpenSource/Conformance/compose not found");
    dir = parent;
  }
}

function corpus(): Json {
  const doc = JSON.parse(readFileSync(join(corpusDir(), "result.json"), "utf-8")) as Json;
  assert.equal(doc["version"], 1, "result.json: version");
  return doc;
}

function section(doc: Json, name: string): Json[] {
  const raw = doc[name];
  const rows = Array.isArray(raw) ? raw : (raw as Json | undefined)?.["cases"];
  assert.ok(Array.isArray(rows) && rows.length > 0, `result.json: ${name} must be a non-empty case array`);
  return rows as Json[];
}

test("compose: the result vocabulary agrees with the corpus", () => {
  assert.deepEqual([...COMPOSE_RESULTS], corpus()["vocabulary"]);
});

test("compose: the result-fidelity table agrees with the corpus", () => {
  const fidelity = corpus()["resultFidelity"] as Json;
  const declared = Object.entries(fidelity).filter(([key]) => !key.startsWith("_"));
  assert.ok(declared.length > 0, "result.json: resultFidelity names no renderers");
  for (const [renderer, actions] of declared) {
    for (const [action, results] of Object.entries(actions as Json)) {
      assert.deepEqual(
        [...(COMPOSE_RESULT_FIDELITY[renderer]?.[action] ?? [])], results,
        `resultFidelity: ${renderer}.${action}`,
      );
    }
  }
});

test("compose: the permission surface is empty on every action", () => {
  const surface = corpus()["permissionSurface"] as Json;
  const declared = Object.entries(surface)
    .filter(([key, value]) => !key.startsWith("_") && typeof value === "string");
  assert.ok(declared.length > 0, "result.json: permissionSurface names no actions");
  for (const [action, grant] of declared) {
    assert.equal(grant, "none", `permissionSurface: ${action} must need no grant`);
    assert.equal(COMPOSE_PERMISSION_SURFACE[action], "none", `permissionSurface: ${action}`);
  }
  assert.deepEqual([...COMPOSE_FORBIDDEN_PERMISSIONS], surface["forbiddenPermissions"]);
  for (const permission of COMPOSE_FORBIDDEN_PERMISSIONS) {
    assert.ok(
      !Object.values(COMPOSE_PERMISSION_SURFACE).includes(permission),
      `${permission} must never appear in the permission surface`,
    );
  }
});

test("compose: the result ladder agrees with the corpus", () => {
  const doc = corpus();
  const fidelity = doc["resultFidelity"] as Json;
  for (const testCase of section(doc, "cases")) {
    const given = testCase["given"] as Json;
    const expect = testCase["expect"] as Json;
    const name = testCase["name"] as string;

    const outcome = composeOutcome({
      renderer: given["renderer"] as string,
      action: given["action"] as string,
      composerResult: given["composerResult"] as string | null,
      launched: given["launched"] as boolean | undefined,
      isHtml: given["isHtml"] as boolean | undefined,
    });

    if (expect["error"] !== undefined && expect["error"] !== null) {
      assert.equal(outcome.error, expect["error"], `${name}: error`);
      continue;
    }
    assert.equal(outcome.error, expect["error"] ?? null, `${name}: error`);
    assert.equal(outcome.result, expect["result"], `${name}: result`);
    if ("isHtml" in expect) assert.equal(outcome.isHtml, expect["isHtml"], `${name}: isHtml`);
    assert.ok(
      (fidelity[given["renderer"] as string][given["action"] as string] as string[]).includes(outcome.result as string),
      `${name}: ${outcome.result} is outside ${given["renderer"]}.${given["action"]}'s fidelity list`,
    );
  }
});

test("compose: the capability disclosure agrees with the corpus", () => {
  for (const testCase of section(corpus(), "capabilities")) {
    const given = testCase["given"] as Json;
    const expect = testCase["expect"] as Json;
    const name = testCase["name"] as string;
    const capabilities = composeCapabilities({
      renderer: given["renderer"] as string,
      canText: given["canText"] as boolean | undefined,
      canMail: given["canMail"] as boolean | undefined,
      smsResolver: given["smsResolver"] as string | undefined,
      mailResolver: given["mailResolver"] as string | undefined,
    });
    assert.equal(capabilities.sms, expect["sms"], `${name}: sms`);
    assert.equal(capabilities.mail, expect["mail"], `${name}: mail`);
    if ("defaultMailClient" in expect) {
      assert.equal(capabilities.defaultMailClient, expect["defaultMailClient"], `${name}: defaultMailClient`);
    }
  }
});

test("compose: the recipient cap agrees with the corpus", () => {
  for (const testCase of section(corpus(), "recipients")) {
    const given = testCase["given"] as Json;
    const expect = testCase["expect"] as Json;
    const count = given["count"] !== undefined
      ? (given["count"] as number)
      : Number(given["to"] ?? 0) + Number(given["cc"] ?? 0) + Number(given["bcc"] ?? 0);
    const decision = composeRecipientDecision(count, given["cap"] as number);
    assert.equal(decision.runs, expect["runs"], `${testCase["name"]}: runs`);
    assert.equal(decision.error, expect["error"] ?? null, `${testCase["name"]}: error`);
  }
});

test("compose: the attachment rule agrees with the corpus", () => {
  for (const testCase of section(corpus(), "attachments")) {
    const given = testCase["given"] as Json;
    const expect = testCase["expect"] as Json;
    const decision = composeAttachmentDecision({
      renderer: given["renderer"] as string,
      path: given["path"] as string,
      insideRoots: given["insideRoots"] as boolean | undefined,
      exists: given["exists"] as boolean | undefined,
    });
    assert.equal(decision.runs, expect["runs"], `${testCase["name"]}: runs`);
    assert.equal(decision.error, expect["error"] ?? null, `${testCase["name"]}: error`);
  }
});
